File: 프로그래머스/lv3/43162. 네트워크/네트워크.py
from collections import deque

ans = 0

def bfs(n, graph, computers):
    global ans
    
    visited = [0] * n
    
    for i in range(n):
        q = deque()
        
        if visited[i]==0:
            visited[i]=1
            ans+=1
            q.append(graph[i])
            while q:
                v = q.popleft()
                for j in range(len(v)):        
                    if visited[v[j]]==0:
                        visited[v[j]]=1
                        q.append(graph[v[j]])
    
def solution(n, computers):
    graph = [[] for _ in range(n)]
    
    for i in range(n):
        for j in range(n):
            if computers[i][j]==1 and i!=j:
                graph[i].append(j)
                
    print(graph)
    bfs(n, graph, computers)
    
    answer = ans
    return answer

File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    
    public int solution(int n, int[][] computers) {
        int answer = 0;
        
        List<List<Integer>> graph = new ArrayList<>();
        for (int i=0; i<n; i++){
            graph.add(new ArrayList<Integer>());
        }
        
        for(int i=0; i<n;i++){
            for(int j=0; j<n; j++) {
                if(i==j) continue;
                
                if(computers[i][j]==1){
                    graph.get(i).add(j);
                    graph.get(j).add(i);
                }
            }
        }
        
        // 방문 여부 저장
        boolean[] visited = new boolean[n];
        
        Queue<Integer> q = new LinkedList<>();
        
        int x;
        for(int i=0; i<n; i++){
            if(visited[i]) continue;
            
            answer++;
            
            visited[i] = true;
            q.add(i);
            while(q.size()>0){
                x = q.poll();
                
                for(Integer to: graph.get(x)){
                    if(visited[to]) continue;
                    
                    visited[to] = true;
                    q.add(to);
                }
            }
            q.clear();
        }

        return answer;
    }
}